src/nodes/statements.ts

```diff
@@ -71,17 +71,48 @@ export class CEmptyStatement {
 }
 
 @CodeTemplate(`
-{destructors}
-return {expression};
+{#if returnTypeAndVar}
+    {
+        {returnTypeAndVar} = {expression};
+        {destructors}
+        return {returnTemp};
+    }
+{#else}
+    {destructors}
+    return {expression};
+{/if}
 `, ts.SyntaxKind.ReturnStatement)
 export class CReturnStatement extends CTemplateBase {
     public expression: CExpression;
     public destructors: CVariableDestructors;
     public retVarName: string = null;
+    public returnTypeAndVar: string = null;
+    public returnTemp: string = null;
     public closureParams: { name: string, value: CExpression }[] = [];
+    doCheckVarNeeded(scope: IScope, node: ts.ReturnStatement): boolean {
+        let s = scope.root.memoryManager.getDestructorsForScope(node);
+        let result = false;
+        for (let e of s) {
+            result = result || new RegExp('\\W' + e.varName + '\\W').test(' ' + node.getFullText() + ' ');
```

Comment:
I am afraid relying on a text match here is too fragile and error-prone.

return statement can be huge, for example something like `return function() { /* ... 200 lines of code here ... */ }`, or something like `return translate("Lorem ipsum dolor ... 5000 characters of random text here")`... it can contain any imaginable text and it will not mean anything.

the right way would be to do something similar to what we do in [memory.ts](https://github.com/andrei-markeev/ts2c/blob/master/src/memory.ts) i.e. properly analyse the return statement and if a subkey of an object that was scheduled to be disposed in this scope escapes into the upper function, then we need to change the return statement to use the temporary variable.